// openskill/dist/index.js
var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));

// src/constants.ts
var builder = (options) => {
  const { z = 3, mu = 25, epsilon = 0.1, alpha = 1, target = 0 } = options;
  const { tau = 25 / 300, sigma = 25 / 3, beta = 25 / 6, limitSigma = false } = options;
  const { balance = false, kappa = 1e-4 } = options;
  const betaSq = beta ** 2;
  return {
    SIGMA: sigma,
    MU: mu,
    EPSILON: epsilon,
    TWOBETASQ: 2 * betaSq,
    BETA: beta,
    BETASQ: betaSq,
    Z: z,
    ALPHA: alpha,
    TARGET: target,
    TAU: tau,
    LIMIT_SIGMA: limitSigma,
    BALANCE: balance,
    KAPPA: kappa
  };
};
var constants_default = builder;

// src/rating.ts
var rating = (init, options = {}) => {
  const { MU: mu, SIGMA: sigma } = constants_default(__spreadValues(__spreadValues({}, options), init));
  return { mu, sigma };
};
var rating_default = rating;

// src/rate.ts
import { sortBy, identity, range } from "ramda";
import { unwind } from "sort-unwind";

// src/margin.ts
var marginAmp = (absDiff, margin) => Math.log1p(Math.max(0, absDiff - margin));
var marginFactor = (scores, i, margin) => {
  const n = scores.length;
  if (n <= 1) return 1;
  const si = scores[i];
  const sum2 = scores.reduce((acc, sj, j) => j === i ? acc : acc + marginAmp(Math.abs(si - sj), margin), 0);
  return 1 + sum2 / (n - 1);
};

// src/util.ts
import { zip } from "ramda";
var sum = (a, b) => a + b;
var rankings = (teams, rank = []) => {
  const teamScores = teams.map((_, i) => {
    var _a;
    return (_a = rank[i]) != null ? _a : i;
  });
  const outRank = new Array(teams.length);
  let s = 0;
  for (let j = 0; j < teamScores.length; j += 1) {
    if (j > 0 && teamScores[j - 1] < teamScores[j]) {
      s = j;
    }
    outRank[j] = s;
  }
  return outRank;
};
var teamRating = (options) => (game) => {
  const { Z, ALPHA, TARGET, BALANCE, KAPPA } = constants_default(options);
  const rank = rankings(game, options.rank);
  return game.map((team, i) => {
    if (!BALANCE) {
      const { mu: muSum2, sigSq: sigSq2 } = team.reduce(
        (a, { mu, sigma }) => {
          a.mu += mu;
          a.sigSq += sigma * sigma;
          return a;
        },
        { mu: 0, sigSq: 0 }
      );
      return [muSum2, sigSq2, team, rank[i]];
    }
    const ordinals = team.map(({ mu, sigma }) => TARGET + ALPHA * (mu - Z * sigma));
    const maxOrdinal = Math.max(...ordinals);
    const denom = maxOrdinal + KAPPA;
    const { mu: muSum, sigSq } = team.reduce(
      (a, { mu, sigma }, j) => {
        const factor = 1 + (maxOrdinal - ordinals[j]) / denom;
        a.mu += mu * factor;
        const s = sigma * factor;
        a.sigSq += s * s;
        return a;
      },
      { mu: 0, sigSq: 0 }
    );
    return [muSum, sigSq, team, rank[i]];
  });
};
var utilC = (options) => {
  const { BETASQ } = constants_default(options);
  return (teamRatings) => Math.sqrt(teamRatings.map(([_teamMu, teamSigmaSq, _team, _rank]) => teamSigmaSq + BETASQ).reduce(sum, 0));
};
var utilSumQ = (teamRatings, c) => teamRatings.map(
  ([_qMu, _qSigmaSq, _qTeam, qRank]) => teamRatings.filter(([_iMu, _iSigmaSq, _iTeam, iRank]) => iRank >= qRank).map(([iMu, _iSigmaSq, _iTeam, _iRank]) => Math.exp(iMu / c)).reduce(sum, 0)
);
var utilA = (teamRatings) => teamRatings.map(
  ([_iMu, _iSigmaSq, _iTeam, iRank]) => teamRatings.filter(([_qMu, _qSigmaSq, _qTeam, qRank]) => iRank === qRank).length
);
var defaultGamma = (c, _k, _mu, sigmaSq, _team, _qRank) => Math.sqrt(sigmaSq) / c;
var gamma = (options) => {
  var _a;
  return (_a = options.gamma) != null ? _a : defaultGamma;
};
var util_default = (options) => ({
  utilC: utilC(options),
  teamRating: teamRating(options),
  gamma: gamma(options)
});

// src/models/plackett-luce.ts
var model = (game, options = {}) => {
  const { KAPPA } = constants_default(options);
  const { utilC: utilC2, teamRating: teamRating2, gamma: gamma2 } = util_default(options);
  const teamRatings = teamRating2(game);
  const c = utilC2(teamRatings);
  const sumQ = utilSumQ(teamRatings, c);
  const a = utilA(teamRatings);
  return teamRatings.map((iTeamRating, i) => {
    const [iMu, iSigmaSq, iTeam, iRank] = iTeamRating;
    const iMuOverCe = Math.exp(iMu / c);
    const { omega: omegaSum, delta: deltaSum } = teamRatings.reduce(
      (acc, [_qMu, _qSigmaSq, _qTeam, qRank], q) => {
        if (qRank > iRank) return acc;
        const quotient = iMuOverCe / sumQ[q];
        acc.omega += (i === q ? 1 - quotient : -quotient) / a[q];
        acc.delta += quotient * (1 - quotient) / a[q];
        return acc;
      },
      { omega: 0, delta: 0 }
    );
    const iGamma = gamma2(c, teamRatings.length, ...iTeamRating);
    const iOmega = omegaSum * (iSigmaSq / c);
    const iDelta = deltaSum * (iSigmaSq / c ** 2) * iGamma;
    return iTeam.map(({ mu, sigma }) => ({
      mu: mu + sigma ** 2 / iSigmaSq * iOmega,
      sigma: sigma * Math.sqrt(Math.max(1 - sigma ** 2 / iSigmaSq * iDelta, KAPPA))
    }));
  });
};
var plackett_luce_default = model;

// src/statistics.ts
import erf from "@stdlib/math-base-special-erf";
import quantile from "@stdlib/stats-base-dists-normal-quantile";
var SQRT_2 = Math.sqrt(2);
var SQRT_TAU = Math.sqrt(2 * Math.PI);
var phiMajor = (x) => 0.5 * (1 + erf(x / SQRT_2));
var phiMajorInverse = (x) => quantile(x, 0, 1);

// src/rate.ts
var rate = (teams, options = {}) => {
  var _a, _b, _c;
  const { LIMIT_SIGMA, TAU } = constants_default(options);
  const { model: model2 = plackett_luce_default } = options;
  const tauSquared = TAU * TAU;
  const processedTeams = teams.map(
    (team) => team.map((p) => __spreadProps(__spreadValues({}, p), {
      sigma: Math.sqrt(p.sigma * p.sigma + tauSquared)
    }))
  );
  const rank = (_c = (_b = options.rank) != null ? _b : (_a = options.score) == null ? void 0 : _a.map((points) => -points)) != null ? _c : range(1, teams.length + 1);
  const [orderedTeams, tenet] = unwind(rank, processedTeams);
  const newRatings = model2(orderedTeams, __spreadProps(__spreadValues({}, options), {
    rank: sortBy(identity, rank)
  }));
  let [reorderedTeams] = unwind(tenet, newRatings);
  if (options.score && options.margin) {
    const scores = options.score;
    const margin = options.margin;
    reorderedTeams = reorderedTeams.map((team, i) => {
      const factor = marginFactor(scores, i, margin);
      return team.map((p, k) => {
        const priorMu = processedTeams[i][k].mu;
        return { mu: priorMu + factor * (p.mu - priorMu), sigma: p.sigma };
      });
    });
  }
  if (TAU && LIMIT_SIGMA) {
    reorderedTeams = reorderedTeams.map(
      (team, i) => team.map((p, j) => __spreadProps(__spreadValues({}, p), {
        sigma: Math.min(p.sigma, teams[i][j].sigma)
      }))
    );
  }
  return reorderedTeams;
};
var rate_default = rate;

// src/ordinal.ts
var ordinal = (rating2, options = {}) => {
  const { sigma, mu } = rating2;
  const { Z, ALPHA, TARGET } = constants_default(options);
  return TARGET + ALPHA * (mu - Z * sigma);
};
var ordinal_default = ordinal;

// src/predict-win.ts
import { reduce } from "ramda";
function* upperTrianglePairs(n) {
  for (let i = 0; i < n; i += 1) {
    for (let j = i + 1; j < n; j += 1) {
      yield [i, j];
    }
  }
}
var predictWin = (teams, options = {}) => {
  const { teamRating: teamRating2 } = util_default(options);
  const { BETASQ } = constants_default(options);
  const teamRatings = teamRating2(teams);
  const n = teams.length;
  const denom = n * (n - 1) / 2;
  const wins = reduce(
    (acc, [i, j]) => {
      const [muA, sigmaSqA] = teamRatings[i];
      const [muB, sigmaSqB] = teamRatings[j];
      const p = phiMajor((muA - muB) / Math.sqrt(n * BETASQ + sigmaSqA + sigmaSqB));
      acc[i] += p;
      acc[j] += 1 - p;
      return acc;
    },
    new Array(n).fill(0),
    upperTrianglePairs(n)
  );
  return wins.map((w) => w / denom);
};
var predict_win_default = predictWin;

// src/predict-draw.ts
import { filter, head, map, range as range2, reduce as reduce2, xprod } from "ramda";
var predictDraw = (teams, options = {}) => {
  const { teamRating: teamRating2 } = util_default(options);
  const { BETASQ, BETA } = constants_default(options);
  const totalPlayerCount = reduce2((acc, t) => acc + t.length, 0, teams);
  const drawProbability = 1 / totalPlayerCount;
  const drawMargin = Math.sqrt(totalPlayerCount) * BETA * phiMajorInverse((1 + drawProbability) / 2);
  const teamRatings = map((team) => head(teamRating2([team])), teams);
  const idx = range2(0, teamRatings.length);
  const pairs = filter(([i, j]) => i < j, xprod(idx, idx));
  const { total, count } = reduce2(
    (acc, [i, j]) => {
      const [muA, sigmaSqA] = teamRatings[i];
      const [muB, sigmaSqB] = teamRatings[j];
      const sharedDenom = Math.sqrt(totalPlayerCount * BETASQ + sigmaSqA + sigmaSqB);
      acc.total += phiMajor((drawMargin - muA + muB) / sharedDenom) - phiMajor((muB - muA - drawMargin) / sharedDenom);
      acc.count += 1;
      return acc;
    },
    { total: 0, count: 0 },
    pairs
  );
  return total / count;
};
var predict_draw_default = predictDraw;

// src/predict-rank.ts
var predictRank = (teams, options = {}) => {
  const winProbs = predict_win_default(teams, options);
  return winProbs.map((prob, _, all) => [1 + all.filter((q) => q > prob).length, prob]);
};
var predict_rank_default = predictRank;
export {
  ordinal_default as ordinal,
  predict_draw_default as predictDraw,
  predict_rank_default as predictRank,
  predict_win_default as predictWin,
  rate_default as rate,
  rating_default as rating
};
