// openskill/dist/predict-draw.js
// src/predict-draw.ts
import { filter, head, map, range, reduce, xprod } from "ramda";

// src/constants.ts
var builder = (options) => {
  const { z = 3, mu = 25, epsilon = 0.1, alpha = 1, target = 0 } = options;
  const { tau = 25 / 300, sigma = 25 / 3, beta = 25 / 6, limitSigma = false } = options;
  const { balance = false, kappa = 1e-4 } = options;
  const betaSq = beta ** 2;
  return {
    SIGMA: sigma,
    MU: mu,
    EPSILON: epsilon,
    TWOBETASQ: 2 * betaSq,
    BETA: beta,
    BETASQ: betaSq,
    Z: z,
    ALPHA: alpha,
    TARGET: target,
    TAU: tau,
    LIMIT_SIGMA: limitSigma,
    BALANCE: balance,
    KAPPA: kappa
  };
};
var constants_default = builder;

// src/util.ts
import { zip } from "ramda";
var sum = (a, b) => a + b;
var rankings = (teams, rank = []) => {
  const teamScores = teams.map((_, i) => {
    var _a;
    return (_a = rank[i]) != null ? _a : i;
  });
  const outRank = new Array(teams.length);
  let s = 0;
  for (let j = 0; j < teamScores.length; j += 1) {
    if (j > 0 && teamScores[j - 1] < teamScores[j]) {
      s = j;
    }
    outRank[j] = s;
  }
  return outRank;
};
var teamRating = (options) => (game) => {
  const { Z, ALPHA, TARGET, BALANCE, KAPPA } = constants_default(options);
  const rank = rankings(game, options.rank);
  return game.map((team, i) => {
    if (!BALANCE) {
      const { mu: muSum2, sigSq: sigSq2 } = team.reduce(
        (a, { mu, sigma }) => {
          a.mu += mu;
          a.sigSq += sigma * sigma;
          return a;
        },
        { mu: 0, sigSq: 0 }
      );
      return [muSum2, sigSq2, team, rank[i]];
    }
    const ordinals = team.map(({ mu, sigma }) => TARGET + ALPHA * (mu - Z * sigma));
    const maxOrdinal = Math.max(...ordinals);
    const denom = maxOrdinal + KAPPA;
    const { mu: muSum, sigSq } = team.reduce(
      (a, { mu, sigma }, j) => {
        const factor = 1 + (maxOrdinal - ordinals[j]) / denom;
        a.mu += mu * factor;
        const s = sigma * factor;
        a.sigSq += s * s;
        return a;
      },
      { mu: 0, sigSq: 0 }
    );
    return [muSum, sigSq, team, rank[i]];
  });
};
var utilC = (options) => {
  const { BETASQ } = constants_default(options);
  return (teamRatings) => Math.sqrt(teamRatings.map(([_teamMu, teamSigmaSq, _team, _rank]) => teamSigmaSq + BETASQ).reduce(sum, 0));
};
var defaultGamma = (c, _k, _mu, sigmaSq, _team, _qRank) => Math.sqrt(sigmaSq) / c;
var gamma = (options) => {
  var _a;
  return (_a = options.gamma) != null ? _a : defaultGamma;
};
var util_default = (options) => ({
  utilC: utilC(options),
  teamRating: teamRating(options),
  gamma: gamma(options)
});

// src/statistics.ts
import erf from "@stdlib/math-base-special-erf";
import quantile from "@stdlib/stats-base-dists-normal-quantile";
var SQRT_2 = Math.sqrt(2);
var SQRT_TAU = Math.sqrt(2 * Math.PI);
var phiMajor = (x) => 0.5 * (1 + erf(x / SQRT_2));
var phiMajorInverse = (x) => quantile(x, 0, 1);

// src/predict-draw.ts
var predictDraw = (teams, options = {}) => {
  const { teamRating: teamRating2 } = util_default(options);
  const { BETASQ, BETA } = constants_default(options);
  const totalPlayerCount = reduce((acc, t) => acc + t.length, 0, teams);
  const drawProbability = 1 / totalPlayerCount;
  const drawMargin = Math.sqrt(totalPlayerCount) * BETA * phiMajorInverse((1 + drawProbability) / 2);
  const teamRatings = map((team) => head(teamRating2([team])), teams);
  const idx = range(0, teamRatings.length);
  const pairs = filter(([i, j]) => i < j, xprod(idx, idx));
  const { total, count } = reduce(
    (acc, [i, j]) => {
      const [muA, sigmaSqA] = teamRatings[i];
      const [muB, sigmaSqB] = teamRatings[j];
      const sharedDenom = Math.sqrt(totalPlayerCount * BETASQ + sigmaSqA + sigmaSqB);
      acc.total += phiMajor((drawMargin - muA + muB) / sharedDenom) - phiMajor((muB - muA - drawMargin) / sharedDenom);
      acc.count += 1;
      return acc;
    },
    { total: 0, count: 0 },
    pairs
  );
  return total / count;
};
var predict_draw_default = predictDraw;
export {
  predict_draw_default as default
};
