// openskill/dist/models/thurstone-mosteller-part.js
// src/models/thurstone-mosteller-part.ts
import { zip as zip2 } from "ramda";

// src/util.ts
import { zip } from "ramda";

// src/constants.ts
var builder = (options) => {
  const { z = 3, mu = 25, epsilon = 0.1, alpha = 1, target = 0 } = options;
  const { tau = 25 / 300, sigma = 25 / 3, beta = 25 / 6, limitSigma = false } = options;
  const { balance = false, kappa = 1e-4 } = options;
  const betaSq = beta ** 2;
  return {
    SIGMA: sigma,
    MU: mu,
    EPSILON: epsilon,
    TWOBETASQ: 2 * betaSq,
    BETA: beta,
    BETASQ: betaSq,
    Z: z,
    ALPHA: alpha,
    TARGET: target,
    TAU: tau,
    LIMIT_SIGMA: limitSigma,
    BALANCE: balance,
    KAPPA: kappa
  };
};
var constants_default = builder;

// src/util.ts
var sum = (a, b) => a + b;
var rankings = (teams, rank = []) => {
  const teamScores = teams.map((_, i) => {
    var _a;
    return (_a = rank[i]) != null ? _a : i;
  });
  const outRank = new Array(teams.length);
  let s = 0;
  for (let j = 0; j < teamScores.length; j += 1) {
    if (j > 0 && teamScores[j - 1] < teamScores[j]) {
      s = j;
    }
    outRank[j] = s;
  }
  return outRank;
};
var teamRating = (options) => (game) => {
  const { Z, ALPHA, TARGET, BALANCE, KAPPA } = constants_default(options);
  const rank = rankings(game, options.rank);
  return game.map((team, i) => {
    if (!BALANCE) {
      const { mu: muSum2, sigSq: sigSq2 } = team.reduce(
        (a, { mu, sigma }) => {
          a.mu += mu;
          a.sigSq += sigma * sigma;
          return a;
        },
        { mu: 0, sigSq: 0 }
      );
      return [muSum2, sigSq2, team, rank[i]];
    }
    const ordinals = team.map(({ mu, sigma }) => TARGET + ALPHA * (mu - Z * sigma));
    const maxOrdinal = Math.max(...ordinals);
    const denom = maxOrdinal + KAPPA;
    const { mu: muSum, sigSq } = team.reduce(
      (a, { mu, sigma }, j) => {
        const factor = 1 + (maxOrdinal - ordinals[j]) / denom;
        a.mu += mu * factor;
        const s = sigma * factor;
        a.sigSq += s * s;
        return a;
      },
      { mu: 0, sigSq: 0 }
    );
    return [muSum, sigSq, team, rank[i]];
  });
};
var ladderPairs = (ranks) => {
  const size = ranks.length;
  const left = [void 0, ...ranks.slice(0, size - 1)];
  const right = [...ranks.slice(1), void 0];
  return zip(left, right).map(([l, r]) => {
    if (l !== void 0 && r !== void 0) return [l, r];
    if (l !== void 0 && r === void 0) return [l];
    if (l === void 0 && r !== void 0) return [r];
    return [];
  });
};
var utilC = (options) => {
  const { BETASQ } = constants_default(options);
  return (teamRatings) => Math.sqrt(teamRatings.map(([_teamMu, teamSigmaSq, _team, _rank]) => teamSigmaSq + BETASQ).reduce(sum, 0));
};
var defaultGamma = (c, _k, _mu, sigmaSq, _team, _qRank) => Math.sqrt(sigmaSq) / c;
var gamma = (options) => {
  var _a;
  return (_a = options.gamma) != null ? _a : defaultGamma;
};
var util_default = (options) => ({
  utilC: utilC(options),
  teamRating: teamRating(options),
  gamma: gamma(options)
});

// src/statistics.ts
import erf from "@stdlib/math-base-special-erf";
import quantile from "@stdlib/stats-base-dists-normal-quantile";
var SQRT_2 = Math.sqrt(2);
var SQRT_TAU = Math.sqrt(2 * Math.PI);
var phiMajor = (x) => 0.5 * (1 + erf(x / SQRT_2));
var phiMinor = (x) => Math.exp(x * x / -2) / SQRT_TAU;
var v = (x, t) => {
  const xt = x - t;
  const denom = phiMajor(xt);
  return denom < Number.EPSILON ? -xt : phiMinor(xt) / denom;
};
var w = (x, t) => {
  const xt = x - t;
  const denom = phiMajor(xt);
  if (denom < Number.EPSILON) {
    return x < 0 ? 1 : 0;
  }
  return v(x, t) * (v(x, t) + xt);
};
var vt = (x, t) => {
  const xx = Math.abs(x);
  const b = phiMajor(t - xx) - phiMajor(-t - xx);
  if (b < 1e-5) {
    if (x < 0) return -x - t;
    return -x + t;
  }
  const a = phiMinor(-t - xx) - phiMinor(t - xx);
  return (x < 0 ? -a : a) / b;
};
var wt = (x, t) => {
  const xx = Math.abs(x);
  const b = phiMajor(t - xx) - phiMajor(-t - xx);
  return b < Number.EPSILON ? 1 : ((t - xx) * phiMinor(t - xx) + (t + xx) * phiMinor(-t - xx)) / b + vt(x, t) * vt(x, t);
};

// src/models/thurstone-mosteller-part.ts
var model = (game, options = {}) => {
  const { TWOBETASQ, EPSILON, KAPPA } = constants_default(options);
  const { teamRating: teamRating2, gamma: gamma2 } = util_default(options);
  const teamRatings = teamRating2(game);
  const adjacentTeams = ladderPairs(teamRatings);
  return zip2(teamRatings, adjacentTeams).map(([iTeamRating, iAdjacents]) => {
    const [iMu, iSigmaSq, iTeam, iRank] = iTeamRating;
    const { omega: iOmega, delta: iDelta } = iAdjacents.reduce(
      (acc, [qMu, qSigmaSq, _qTeam, qRank]) => {
        const ciq = 2 * Math.sqrt(iSigmaSq + qSigmaSq + TWOBETASQ);
        const deltaMu = (iMu - qMu) / ciq;
        const qEta = iSigmaSq / ciq;
        const iGamma = gamma2(ciq, teamRatings.length, ...iTeamRating);
        if (qRank === iRank) {
          acc.omega += qEta * vt(deltaMu, EPSILON / ciq);
          acc.delta += iGamma * qEta / ciq * wt(deltaMu, EPSILON / ciq);
          return acc;
        }
        const sign = qRank > iRank ? 1 : -1;
        acc.omega += sign * qEta * v(sign * deltaMu, EPSILON / ciq);
        acc.delta += iGamma * qEta / ciq * w(sign * deltaMu, EPSILON / ciq);
        return acc;
      },
      { omega: 0, delta: 0 }
    );
    return iTeam.map(({ mu, sigma }) => {
      const sigmaSq = sigma * sigma;
      return {
        mu: mu + sigmaSq / iSigmaSq * iOmega,
        sigma: sigma * Math.sqrt(Math.max(1 - sigmaSq / iSigmaSq * iDelta, KAPPA))
      };
    });
  });
};
var thurstone_mosteller_part_default = model;
export {
  thurstone_mosteller_part_default as default
};
